File: node/sdk/src/service/validate.ts
import { Code, ConnectError } from "@connectrpc/connect";
import { createValidator } from "@bufbuild/protovalidate";
import type { DescMessage, MessageShape } from "@bufbuild/protobuf";

/**
 * Shared protovalidate validator instance for the public {@link validate} helper.
 * Construction is cheap, but reusing one instance avoids repeated compilation
 * of the same rules.
 */
export const validator = createValidator();

/**
 * Validates a response message against its buf.validate proto annotations.
 *
 * On success, returns the message unchanged (typed). On failure, throws a
 * {@link ConnectError} with {@link Code.Internal} and a message matching the
 * shape emitted by the SDK's response-validation interceptor — so propagating
 * the error from a handler produces the same wire response as not calling
 * `validate` at all.
 *
 * Intended use:
 * ```ts
 * return validate(PayOutResponseSchema, { result: { case: "accepted", value: {} } });
 * ```
 *
 * Catch the error to convert it into a domain-level failure (e.g. the `Failed`
 * arm of a `oneof result`) instead of an opaque `Code.Internal`.
 */
export function validate<Desc extends DescMessage>(
  schema: Desc,
  msg: MessageShape<Desc>,
): MessageShape<Desc> {
  const result = validator.validate(schema, msg);
  if (result.kind === "invalid") {
    const details = result.violations
      .map((v) => `${v.field?.toString() ?? ""}: ${v.message}`)
      .join("; ");
    throw new ConnectError(`response validation failed: ${details}`, Code.Internal);
  }
  if (result.kind === "error") {
    throw new ConnectError(
      `response validation error: ${result.error.message}`,
      Code.Internal,
    );
  }
  return msg;
}

File: java/sdk/src/main/java/network/t0/sdk/provider/Validate.java
package network.t0.sdk.provider;

import build.buf.protovalidate.ValidationResult;
import build.buf.protovalidate.exceptions.ValidationException;
import com.google.protobuf.Message;
import network.t0.sdk.common.ValidationUtils;

/**
 * Public helper that lets provider handlers explicitly validate a response
 * message against its {@code protovalidate} constraints before returning it
 * to the SDK.
 *
 * <p>The SDK's {@link ResponseValidationInterceptor} already validates every
 * outgoing response as a safety net, but by then the handler has already
 * returned, so the developer never sees the failure. Calling
 * {@link #check(Message)} surfaces the failure in the same call frame as
 * the handler, where it can be caught, logged, or converted into a domain-level
 * error.
 *
 * <p>Wire behavior is unchanged: if a {@link ResponseValidationException}
 * thrown by this helper propagates out of the handler, the SDK interceptor
 * closes the call with the same {@code Status.INTERNAL} + same description
 * it would produce on its own.
 *
 * <p>Example:
 * <pre>{@code
 * @Override
 * public void payOut(PayoutRequest req, StreamObserver<PayoutResponse> obs) {
 *     PayoutResponse resp = PayoutResponse.newBuilder()
 *             .setAccepted(PayoutResponse.Accepted.getDefaultInstance())
 *             .build();
 *     obs.onNext(Validate.check(resp));
 *     obs.onCompleted();
 * }
 * }</pre>
 */
public final class Validate {

    private Validate() {}

    /**
     * Validates {@code msg} against its proto-level constraints.
     *
     * @param msg the message to validate
     * @param <T> the concrete message type
     * @return {@code msg} unchanged, on success
     * @throws ResponseValidationException if validation fails or the validator itself errors
     */
    public static <T extends Message> T check(T msg) {
        if (msg == null) {
            return null;
        }
        try {
            ValidationResult result = Validators.shared().validate(msg);
            if (!result.isSuccess()) {
                throw new ResponseValidationException(ValidationUtils.formatViolations(result));
            }
            return msg;
        } catch (ValidationException e) {
            throw new ResponseValidationException(e.getMessage(), e);
        }
    }
}
